mypy/stubgen.py

```diff
@@ -892,86 +843,217 @@ def get_qualified_name(o: Expression) -> str:
-        package = importlib.import_module(package_name)
-        yield package.__name__
-        # get the path of the object (needed by pkgutil)
-        path = getattr(package, '__path__', None)
-        if path is None:
-            # object has no path; this means it's either a module inside a package
-            # (and thus no sub-packages), or it could be a C extension package.
-            if is_c_module(package):
-                # This is a C extension module, now get the list of all sub-packages
-                # using the inspect module
-                subpackages = [package.__name__ + "." + name
-                               for name, val in inspect.getmembers(package)
-                               if inspect.ismodule(val)
-                               and val.__name__ == package.__name__ + "." + name]
-                # recursively iterate through the subpackages
-                for submodule in walk_packages(subpackages):
-                    yield submodule
-            # It's a module inside a package.  There's nothing else to walk/yield.
+    if options.packages or options.modules:
+        if options.no_import:
+            py_modules = find_module_paths_using_search(options.modules,
+                                                        options.packages,
+                                                        options.search_path)
+            c_modules = []  # type: List[StubSource]
         else:
-            all_packages = pkgutil.walk_packages(path, prefix=package.__name__ + ".",
-                                                 onerror=lambda r: None)
-            for importer, qualified_name, ispkg in all_packages:
-                yield qualified_name
+            # Using imports is the default, since we can also find C modules.
+            py_modules, c_modules = find_module_paths_using_imports(options.modules,
+                                                                    options.packages,
+                                                                    options.interpreter,
+                                                                    options.pyversion)
+    else:
+        # Use mypy native source collection for files and directories.
+        try:
+            source_list = create_source_list(options.files, mypy_opts)
+        except InvalidSourceList as e:
+            raise SystemExit(str(e))
+        py_modules = [StubSource(m.module, m.path) for m in source_list]
+        c_modules = []
+
+    return py_modules, c_modules
+
+
+def find_module_paths_using_imports(modules: List[str], packages: List[str],
+                                    interpreter: str,
+                                    pyversion: Tuple[int, int],
+                                    quiet: bool = True) -> Tuple[List[StubSource],
+                                                                 List[StubSource]]:
+    """Find path and runtime value of __all__ (if possible) for modules and packages."""
```

Comment:
This docstring doesn't make it clear how this is different from `find_module_paths_using_search`. Can you update the two docstrings to explain the difference (the other uses runtime Python imports while the other one just looks for source files at the file system level).